### DISCUSSION TITLE
Image in QRCode

### DISCUSSION THREAD

**Discussion by aSamu3l** [chillerlan/js-qrcode#4] [Q&A] [answered]
How can I insert an image in the center of the QR-Code?

**Comment by codemasher** [comment]
Hey, the basic idea is to just append an `<image>` node to the SVG output, see https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image

**Reply by aSamu3l** [reply]
What do I have to put in the constructor to add the image?

**Reply by codemasher** [reply]
Nothing :)
If you're following this example, just set the setting `returnAsDomElement` to `true` which allows you to directly modify the returned SVG element.

https://github.com/chillerlan/js-qrcode/blob/2d66a5759533f64903b8ff6018178f8173c45fca/examples/example-browser-svg.html#L31

The DOM example is at the bottom of the page:

https://github.com/chillerlan/js-qrcode/blob/2d66a5759533f64903b8ff6018178f8173c45fca/examples/example-browser-svg.html#L74-L78

**Reply by aSamu3l** [reply]
I want to insert image into the qr-code
![image](https://github.com/chillerlan/js-qrcode/assets/75342214/e1bf524a-639e-49f8-b81a-d91ba8b51828)

**Reply by codemasher** [reply]
Yes, i get that and i told you how to do it - where's the issue?

**Reply by aSamu3l** [reply]
Sorry, but I'm not understanding what I have to change and how I have to put the image and its measurements

**Reply by codemasher** [reply]
That's not an issue of the library though, that is basic Javascript knowledge:

- https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild
- https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement
- https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image

All of this is used throughout the example and i'm sure you will figure out quickly.

**Reply by aSamu3l** [reply]
I downloaded the svg example, but the image is not in the center of the qr-code and the only thing I see is the qrcode with all the colors. I'm trying to understand what I have to do and I'm not very good at javascript and I can't understand exactly how I have to do it. The only thing I know is how javascript works in html but otherwise I know next to nothing.

**Reply by codemasher** [reply]
And you're asking me to do this work for you. I know that's not the answer you're expecting, but this is not StackOverflow. I gave you the hints, but you'll have to figure out for yourself, which in turn increases your Javascript knowledge.

**Reply by aSamu3l** [reply]
I'm not asking you to do this to me. I'm trying to figure out where I should put the image. I also looked for various examples or something similar. But the documentation is not the best and there are values written that do not correspond even with one of the 2 example codes.
I would like to understand if I have to pass the image in base64 or in some other type and if so in which constructor field should I pass the image

**Reply by codemasher** [reply, marked as answer]
You'll add the image *after* rendering - it has nothing to do with the QRcode at all.

```js
// set options
let options = new qrc.QROptions({
	outputInterface   : qrc.QRMarkupSVG,
	returnAsDomElement: true,
	// ... more options ...
});

// render qrcode
let qrcodeElement = (new qrc.QRCode(options)).render(data); // returns a SVG DOM element


// modify the SVG element (nothing of the following has to do anything with the qrcode library)

// create a container to move and scale the logo
let g = document.createElement('g');

// the hard part: math! aka, calculate the proportions of the logo 
// and its position with respect to the supposed output
g.transform = 'translate(16.875 16.875) scale(0.25)'; // values exemplary

// create the logo
let image = document.createElement('image'); 
image.href = './path/to/image.png'; // can be anything, recommended is base64 data URI

// append image to transform container
g.appendChild(image);
// append transform to SGV element
qrcodeElement.appendChild(g);
// append SVG to body (or whatever container is supposed to hold it)
document.body.appendChild(qrcodeElement);
```

For the transform attribute see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/transform

**Reply by aSamu3l** [reply]
Oh, sorry. I thought the image was inserted directly into the qr-code, in fact I wasn't understanding how it worked

**Reply by codemasher** [reply]
That's what i said in my first reply.
